src/app/modules/verification/verification.controller.ts

```diff
@@ -201,6 +203,68 @@ export const handleVerifyOtp: RequestHandler<
     })
 }
 
+/**
+ * NOTE: Exported solely for testing
+ * Handler for otp verification; double checks the submitted otp against the true otp
+ * If the submitted otp is correct,
+ * the signature that was saved will be appended to the response of the form when submitted
+ * @param formId The id of the form which verification is for
+ * @param transactionId The id of the transaction to validate
+ * @param fieldId The id of the field to validate
+ * @returns 200 when the otp is correct and the parameters are valid
+ * @returns 400 when TransactionExpiredError occurs
+ * @returns 400 when MissingHashDataError occurs
+ * @returns 404 when FormNotFoundError occurs
+ * @returns 404 when TransactionNotFoundError occurs
+ * @returns 404 when FieldNotFoundInTransactionError occurs
+ * @returns 422 when OtpExpiredError occurs
+ * @returns 422 when OtpRetryExceededError occurs
+ * @returns 422 when WrongOtpError occurs
+ * @returns 500 when HashingError occurs
+ * @returns 500 when DatabaseError occurs
+ */
+export const _handleOtpVerification: RequestHandler<
+  { transactionId: string; fieldId: string; formId: string },
+  string | ErrorDto,
+  { otp: string }
+> = async (req, res) => {
+  const { transactionId, fieldId, formId } = req.params
+  const { otp } = req.body
+  const logMeta = {
+    action: '_handleOtpVerification',
+    transactionId,
+    fieldId,
+    ...createReqMeta(req),
+  }
+  return FormService.retrieveFormById(formId)
+    .andThen(() => VerificationFactory.verifyOtp(transactionId, fieldId, otp))
+    .map((signedData) => res.status(StatusCodes.OK).json(signedData))
+    .mapErr((error) => {
+      logger.error({
+        message: 'Error verifying OTP',
+        meta: logMeta,
+        error,
+      })
+      const { statusCode, errorMessage } = mapRouteError(error)
+      return res.status(statusCode).json({ message: errorMessage })
+    })
+}
+
+/**
+ * Handler with otp validation for POST /forms/:formId/fieldverifications/:id/fields/:fieldId/otp/verify
+ */
+export const handleOtpVerification = [
+  celebrate({
+    [Segments.BODY]: Joi.object({
+      otp: Joi.string()
+        .required()
+        .regex(/^\d{6}$/)
+        .message('Please enter a valid OTP'),
+    }),
```

Comment:
out of scope but someday we should standardise whether we rely on Express routing or Joi validation to validate route parameters. personally I'd prefer it if all the validation was in one place, which would mean you have `[Segments.PARAMS]` here as well